wasm/spectests/util.go

```diff
@@ -0,0 +1,175 @@
+package spectests
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/tetratelabs/wazero/wasm"
+)
+
+type (
+	Testbase struct {
+		SourceFile string    `json:"source_filename"`
+		Commands   []Command `json:"commands"`
+	}
+	Command struct {
+		CommandType string `json:"type"`
+		Line        int    `json:"line"`
+
+		// Set when type == "module" || "register"
+		Name string `json:"name,omitempty"`
+
+		// Set when type == "module" || "assert_uninstantiable" || "assert_malformed"
+		Filename string `json:"filename,omitempty"`
+
+		// Set when type == "register"
+		As string `json:"as,omitempty"`
+
+		// Set when type == "assert_return" || "action"
+		Action CommandAction      `json:"action,omitempty"`
+		Exps   []CommandActionVal `json:"expected"`
+
+		// Set when type == "assert_malformed"
+		ModuleType string `json:"module_type"`
+	}
+
+	CommandAction struct {
+		ActionType string             `json:"type"`
+		Args       []CommandActionVal `json:"args"`
+
+		// Set when ActionType == "invoke"
+		Field  string `json:"field,omitempty"`
+		Module string `json:"module,omitempty"`
+	}
+
+	CommandActionVal struct {
+		ValType string `json:"type"`
+		Value   string `json:"value"`
+	}
+)
+
+func (c CommandActionVal) String() string {
+	var v string
+	switch c.ValType {
+	case "i32":
+		v = c.Value
+	case "f32":
+		ret, _ := strconv.ParseUint(c.Value, 10, 32)
+		v = fmt.Sprintf("%f", math.Float32frombits(uint32(ret)))
+	case "i64":
+		v = c.Value
+	case "f64":
+		ret, _ := strconv.ParseUint(c.Value, 10, 64)
+		v = fmt.Sprintf("%f", math.Float64frombits(ret))
+	}
+	return fmt.Sprintf("{type: %s, value: %v}", c.ValType, v)
+}
+
+func (c Command) String() string {
+	msg := fmt.Sprintf("line: %d, type: %s", c.Line, c.CommandType)
+	switch c.CommandType {
+	case "register":
+		msg += fmt.Sprintf(", name: %s, as: %s", c.Name, c.As)
+	case "module":
+		if c.Name != "" {
+			msg += fmt.Sprintf(", name: %s, filename: %s", c.Name, c.Filename)
+		} else {
+			msg += fmt.Sprintf(", filename: %s", c.Filename)
+		}
+	case "assert_return", "action":
+		msg += fmt.Sprintf(", action type: %s", c.Action.ActionType)
+		if c.Action.Module != "" {
+			msg += fmt.Sprintf(", module: %s", c.Action.Module)
+		}
+		msg += fmt.Sprintf(", field: %s", c.Action.Field)
+		msg += fmt.Sprintf(", args: %v, expected: %v", c.Action.Args, c.Exps)
+	case "assert_malformed":
+		// TODO:
+	case "assert_trap":
+		// TODO:
+	case "assert_invalid":
+		// TODO:
+	case "assert_exhaustion":
+		// TODO:
+	case "assert_unlinkable":
+		// TODO:
+	case "assert_uninstantiable":
+		// TODO:
+	}
+	return "{" + msg + "}"
+}
+
+func (c Command) GetAssertReturnArgs() []uint64 {
```

Comment:
maybe comment somewhere why all return vals are coerced to uint64.. is this because it is the widest type and we reduce bits as needed upon evaluation?

Reply:
this is not only the place where we have all in uint64 -- Wasm only has int32,int64(unsigned),float32,float64 and we have to represent them in one type and that's why we use uint64 for all types here and elsewhere.

Reply:
would it hurt to use a type for []uint64 where the intent is the same? That would allow consolidation of the intent which is to be the widest type to handle all arg types.